### PR TITLE
New interface to generate/modify Private/Public key for SIDP Users only #1689

### PR DESCRIPTION
**Pull request by anatol-sialitski**

### CODE REVIEW

**Review by rymsha — APPROVED** [review]

**Review by alansemenov — CHANGES_REQUESTED** [review]
UI

1. `PublicKeysGrid` should not be shown when it's empty.
2. `PublicKeysGrid` should be inside the input view with the "Add key" button, otherwise it sits too tight to the "Change password" button above.
3. The "Authentication" section started to look messy, with password and public keys together. Maybe we can separate them with different labels (like Roles and Groups in the Roles & Groups section below)? 
4. If we have "Public Keys" as the label inside input view then we won't need another label (which you made bold). Also change button caption to simply say "Add"

Here's my proposal:
image
image

File: src/main/resources/assets/js/app/view/PublicKeysGrid.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        super(className || '');
+
+        this.getEl().addClass('public-keys-grid');
+
+        const publicKeysHeader = new LabelEl(i18n('field.userKeys.grid.title'));
+        publicKeysHeader.addClass('public-keys-label');
+        this.appendChild(publicKeysHeader);
+
+
+        const header = this.createDivEl('public-keys-row', 'rowgroup');
```
Comment (alansemenov) [inline src/main/resources/assets/js/app/view/PublicKeysGrid.ts:29]: Use a different style for the header element, to be able to apply different styling to the header row.

File: src/main/resources/assets/js/app/view/PublicKeysGrid.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        header.appendChild(this.createDivEl('public-keys-cell', 'columnheader', i18n('field.userKeys.grid.kid.column')));
+        header.appendChild(this.createDivEl('public-keys-cell', 'columnheader', i18n('field.userKeys.grid.label.column')));
+        header.appendChild(this.createDivEl('public-keys-cell', 'columnheader', i18n('field.userKeys.grid.creationTime.column')));
+        header.appendChild(this.createDivEl('public-keys-cell', 'columnheader', i18n('field.userKeys.grid.actions.column')));
```
Comment (alansemenov) [inline src/main/resources/assets/js/app/view/PublicKeysGrid.ts:33]: This needs some refactoring:
* Add separate methods that create a regular cell and a header cell:

```Typescript
private createCell(html: string, role?: string): DivEl {
    return this.createDivEl('public-keys-cell', role || 'cell', html);
}

private createHeaderCell(html: string): DivEl {
    return this.createCell(html, 'columnheader');
}
```
* Create separate methods for creating a header and a regular row:

```Typescript
private createHeader(): DivEl {
      const header = this.createDivEl('public-keys-header', 'gridheader');
      header.appendChild(this.createHeaderCell(i18n('field.userKeys.grid.kid.column')));
      header.appendChild(this.createHeaderCell(i18n('field.userKeys.grid.label.column')));
      header.appendChild(this.createHeaderCell(i18n('field.userKeys.grid.creationTime.column')));
      header.appendChild(this.createHeaderCell(i18n('field.userKeys.grid.actions.column')));

      return header;
}

private createDataRow(publicKey: PublicKey): DivEl {
      const row = this.createDivEl('public-keys-row', 'rowgroup');
      row.appendChild(this.createCell(publicKey.getKid()));
      row.appendChild(this.createCell(publicKey.getLabel()));

      const creationTime = new Date(Date.parse(publicKey.getCreationTime()));
      row.appendChild(this.createCell(DateHelper.formatDateTime(creationTime)));

      return row;
}
```

* Change your code to call these new methods when needed.

File: src/main/resources/assets/js/app/view/PublicKeysGrid.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    extends DivEl {
+
+    private readonly tbody: DivEl;
+
+    private removeCallback: () => void;
+
+    constructor(className ?: string) {
+        super(className || '');
+
+        this.getEl().addClass('public-keys-grid');
```
Comment (alansemenov) [inline src/main/resources/assets/js/app/view/PublicKeysGrid.ts:22]: Move this CSS class to a private static property
```Typescript
private static GRID_CSS_STYLE = 'public-keys-grid';
```

then use it as a prefix for class names of other UI elements:
```Typescript
this.getEl().addClass(PublicKeysGrid.GRID_CSS_STYLE);
...
this.createDivEl(`${PublicKeysGrid.GRID_CSS_STYLE}-header`, 'gridheader');
....
this.createDivEl(`${PublicKeysGrid.GRID_CSS_STYLE}-header-cell`, 'headercell');
etc
```

**Review by alansemenov — CHANGES_REQUESTED** [review]

**Comment by alansemenov** [comment]
image

In this dialog:
* Change header to "New Public Key" (I don't think you can add several here? otherwise the UI should look completely different)
* Change caption of the main action button to "Generate"
* There should be a hint/helptext under the field which explains format of the expected value. I tried to enter random values and it never worked for me - the dialog just closed without anything happening. There was an error message in the browser console (which user will never see).
* There should be a validation in this form which validates entered value and gives an error message without closing the dialog.

**Review by alansemenov — APPROVED** [review]

**Pull request merged**